planner/core/expression_rewriter.go

```diff
@@ -762,21 +762,27 @@ func (er *expressionRewriter) handleInSubquery(ctx context.Context, v *ast.Patte
 		er.err = expression.ErrOperandColumns.GenWithStackByArgs(lLen)
 		return v, true
 	}
-	var rexpr expression.Expression
-	if np.Schema().Len() == 1 {
-		rexpr = np.Schema().Columns[0]
-		rCol := rexpr.(*expression.Column)
-		// For AntiSemiJoin/LeftOuterSemiJoin/AntiLeftOuterSemiJoin, we cannot treat `in` expression as
-		// normal column equal condition, so we specially mark the inner operand here.
-		if v.Not || asScalar {
+	// For AntiSemiJoin/LeftOuterSemiJoin/AntiLeftOuterSemiJoin, we cannot treat `in` expression as
+	// normal column equal condition, so we specially mark the inner operand here.
+	if v.Not || asScalar {
+		for i, rCol := range np.Schema().Columns {
+			larg := lexpr
+			lSf, isSf := lexpr.(*expression.ScalarFunction)
+			if isSf {
+				larg = lSf.GetArgs()[i]
+			}
 			rCol.InOperand = true
-			// If both input columns of `in` expression are not null, we can treat the expression
-			// as normal column equal condition instead.
-			lCol, ok := lexpr.(*expression.Column)
-			if ok && mysql.HasNotNullFlag(lCol.GetType().Flag) && mysql.HasNotNullFlag(rCol.GetType().Flag) {
+			// If both input columns of `!= all / = any` expression are not null, we can treat the expression
+			// as normal column equal condition.
+			if mysql.HasNotNullFlag(rCol.GetType().Flag) && expression.NotNull(larg) {
```

Comment:
The conditions to get rid of `in` are limited;
```
mysql> explain select  (t.a in (select s.a from s)) **is true** from t;
+-------------------------+----------+-----------+-----------------------------------------------------------------------------------------+
| id                      | count    | task      | operator info                                                                           |
+-------------------------+----------+-----------+-----------------------------------------------------------------------------------------+
| Projection_6            | 10000.00 | root      | istrue(Column#8)                                                                        |
| └─HashLeftJoin_7        | 10000.00 | root      | CARTESIAN left outer semi join, inner:TableReader_11, other cond:eq(Column#1, Column#4) |
|   ├─TableReader_9       | 10000.00 | root      | data:TableScan_8                                                                        |
|   │ └─TableScan_8       | 10000.00 | cop[tikv] | table:t, range:[-inf,+inf], keep order:false, stats:pseudo                              |
|   └─TableReader_11      | 3.00     | root      | data:TableScan_10                                                                       |
|     └─TableScan_10      | 3.00     | cop[tikv] | table:s, range:[-inf,+inf], keep order:false, stats:pseudo                              |
+-------------------------+----------+-----------+-----------------------------------------------------------------------------------------+
6 rows in set (0.00 sec)
```
() is true above the subquery can also get rid of `in`, so that it can be converted to a inner join.

Reply:
After considering scalar function, `eq(Column#1, Column#4)`  is treated as euqal condition now.But how can it be converted to inner join?

Reply:
I find that `() is true` is the only special case where the `in subquery` can be converted to `exists subquery` . In other most cases, the `in subquery` should return null, false or true in the projection clause.
So we may ignore the optimization for `() is true` here.